gitpod-io/gitpod
# Add tracing to Werft jobs

## Description

This instruments our Werft jobs using the OpenTelemetry SDK for Node. It is configured to send traces directly to Honeycomb to the [werft dataset](https://ui.honeycomb.io/gitpod/datasets/werft) - anyone with a @gitpod.io will have access to the traces by logging into Honeycomb [here](https://ui.honeycomb.io/)

This promotes the current global `werft` object to a class (we need to carry state, the root span) and instruments each of the methods to also produce relevant spans. This means that all job definitions will continue to use the Werft concepts of jobs, phases, and slices.

Unfortunately we're referencing a global werft instance in a lot of places - this was fine before as it didn't have any state - but now we need a specific Werft instance. To avoid having to refactor the code I have decided to expose a singleton Werft instance and put in checks to ensure that we never try to create more than one, and if you try to access the global instance before it has been instantiated it will also error. This is not especially pretty, but I found it to be better than attempt to refactor everything.

An example trace of a successful build job can be see below (notice that all job configuration values are added as span attributes, this specific one had "with clean slate" set).

![Screenshot 2021-11-03 at 12 41 32](https://user-images.githubusercontent.com/83561/140054062-ecf2f8e3-9447-4c08-b960-41c52026cf99.png)

An example of a job that failed is shown below. Notice how the `status_code` is set to `2` and `status_message` contains the relevant error - both of these attributes are standard from OpenTelemetry. 

![Screenshot 2021-11-03 at 12 04 59](https://user-images.githubusercontent.com/83561/140052126-c1c251cb-dc69-4aa1-91c7-9e08a90a79e9.png)

## Related Issue(s)

This Fixes gitpod-io/ops#180 - we decided to go with traces rather than metrics as the traces are more rich in structure and will help us debug performance regressions better than metrics, while still giving us the ability to measure success rate due to Honeycomb's query language for traces. Additionally to push metrics from our Werft jobs to our monitoring stack we would need to set up a push gateway which was perceived to be more work.

Specifically with this PR we will have spans for each phases. We care about the spans where `werft.phase.name` is set to `deploy`. We care about the success rate of these spans as our goal is to have a success rate of 97% for now.

This can be measure using a neat trick in Honeycomb (see [blog post](https://www.honeycomb.io/blog/level-up-with-derived-columns-two-neat-tricks-that-will-improve-your-observability/)) - we create a derived column which will be set to 100.0 for successful deploy and 0.0 for failed ones. Then our success rate is simply the `AVG` of that column. 

![Screenshot 2021-11-03 at 12 39 29](https://user-images.githubusercontent.com/83561/140053824-bbfdc6ba-1463-4d54-a9c5-630794565124.png)

We can also use the Honeycomb SLO feature very easily (I created an example [here](https://ui.honeycomb.io/gitpod/datasets/werft/slo/tyN74R4Kk9d/Deploy-to-preview-environments)) but as that is an enterprise feature we don't want to depend on that for measuring our success rate, hence the trick above.

## How to test

1. Trigger a job on this branch
2. Once the job has completed look at Honeycomb to see the trace. To get an overview of all traces you can run this query: `COUNT WHERE trace.parent_id does-not-exist GROUP BY name`
3. Click on a the trace to see the trace view or play around with more queries. 

## Release Notes

```release-note
NONE
```

## Documentation

No Gitpod documentation needed.
